logger/redis.go

```diff
@@ -0,0 +1,239 @@
+	return &MagnetometerRedisWrapper{
+		System_time: system_time,
+		Mag_x:       mag_x,
+		Mag_y:       mag_y,
+		Mag_z:       mag_z,
+	}
+}
+
+type Redis struct {
+	DB                 *redis.Client
+	ctx                context.Context
+	maxImuEntries      int
+	maxMagEntries      int
+	maxGnssEntries     int
+	maxGnssAuthEntries int
+}
+
+func NewRedis(maxImuEntries int, maxMagEntries int, maxGnssEntries int, maxGnssAuthEntries int) *Redis {
+	return &Redis{
+		maxImuEntries:      maxImuEntries,
+		maxMagEntries:      maxMagEntries,
+		maxGnssEntries:     maxGnssEntries,
+		maxGnssAuthEntries: maxGnssAuthEntries,
+	}
+}
+
+func (s *Redis) Init() error {
+	fmt.Println("Initializing Redis logger")
+	s.DB = redis.NewClient(&redis.Options{
+		Addr: "localhost:6379", // Use your Redis server address
+	})
+	fmt.Println("Getting context")
+	s.ctx = context.Background()
+
+	// Test the connection with a PING command
+	pong, err := s.DB.Ping(s.ctx).Result()
+	if err != nil {
+		fmt.Printf("Could not connect to Redis: %v\n", err)
+		return fmt.Errorf("ping pong failed")
+	}
+	fmt.Println("Redis connected:", pong)
+	fmt.Println("Redis logger initialized")
+	return nil
+}
+
+func (s *Redis) LogImuData(imudata ImuRedisWrapper) error {
+	// create imu proto
+	newdata := sensordata.ImuData{
+		SystemTime: imudata.System_time.String(),
+		Accelerometer: &sensordata.ImuData_AccelerometerData{
+			X: imudata.Accel.X,
+			Y: imudata.Accel.Y,
+			Z: imudata.Accel.Z,
+		},
+		Gyroscope: &sensordata.ImuData_GyroscopeData{
+			X: imudata.Gyro.X,
+			Y: imudata.Gyro.Y,
+			Z: imudata.Gyro.Z,
+		},
+		Temperature: imudata.Temp,
+		Time:        imudata.Time.String(),
+	}
+	// serialize the data
+	protodata, err := prototext.Marshal(&newdata)
+	if err != nil {
+		return err
+	}
+
+	// Push the JSON data to the Redis list
+	if err := s.DB.LPush(s.ctx, "imu_data", protodata).Err(); err != nil {
+		return err
+	}
+
+	if err := s.DB.LTrim(s.ctx, "imu_data", 0, int64(s.maxImuEntries)).Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Redis) LogMagnetometerData(magdata MagnetometerRedisWrapper) error {
+	// create magnetometer proto
+	newdata := sensordata.MagnetometerData{
+		SystemTime: magdata.System_time.String(),
+		X:          magdata.Mag_x,
+		Y:          magdata.Mag_y,
+		Z:          magdata.Mag_z,
+	}
+	// serialize the data
+	protodata, err := prototext.Marshal(&newdata)
+	if err != nil {
+		return err
+	}
+
+	// Push the JSON data to the Redis list
+	if err := s.DB.LPush(s.ctx, "magnetometer_data", protodata).Err(); err != nil {
+		return err
+	}
+	if err := s.DB.LTrim(s.ctx, "magnetometer_data", 0, int64(s.maxMagEntries)).Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Redis) LogGnssData(gnssdata neom9n.Data) error {
+	// Create gnss proto
+	newdata := sensordata.GnssData{
+		Ttff:             gnssdata.Ttff,
```

Comment:
Correct me if I'm wrong, but I think the best for @derekhive would be to receive the raw data, and as raw as possible. This way we pass him formatted and aggregated data.
@mazar-farran if you will take a look at `HandleUbxMessage` method of gnss, I think passing full and every msg is what will be better here:
https://github.com/Hivemapper/hivemapper-data-logger/blob/add-snr/gnss-controller/device/neom9n/datafeed.go#L226